# Learn X in Y Minutes
## Where X=Pandas

_Inspired by the popular [learnxinyminutes](https://learnxinyminutes.com/)._

Pandas is the shorthand for 'Python and Data Analysis'. It provides a rich set of features for exploring and manipulating data, making it the go-to toolkit for a lot of data scientists. 

This is intended to be an intro or a refresher.

Since Pandas is a Python library, you might want to check out [Python](https://learnxinyminutes.com/docs/python/) first.

---

# Basics
Let's get the hang of Pandas!

In [1]:
# import pandas
import pandas as pd

# 0.22 is the latest ao 01/18
pd.__version__

u'0.22.0'

In [2]:
# Panda's dataframe object is like an SQL table; made up of rows and columns
import random
df = pd.DataFrame([[random.randint(0,9) for i in range(10)] for i in range(5)],
                  index=[i for i in range(5)], 
                  columns=list('abcdefghij'))

df # => a 5*10 matrix/table

a  b  c  d  e  f  g  h  i  j
0  5  7  8  5  0  3  7  9  5  2
1  0  3  5  1  8  4  7  0  5  6
2  6  3  6  3  8  5  4  6  2  2
3  2  0  5  0  2  4  5  1  1  7
4  2  0  8  0  3  1  0  5  1  1

In [3]:
# Add another column
df['grp'] = pd.DataFrame(['a', 'b'] * 5)

In [4]:
# ..headers
df.columns

Index([u'a', u'b', u'c', u'd', u'e', u'f', u'g', u'h', u'i', u'j', u'grp'], dtype='object')

In [5]:
# ..just the first ones
df.head(2)

a  b  c  d  e  f  g  h  i  j grp
0  5  7  8  5  0  3  7  9  5  2   a
1  0  3  5  1  8  4  7  0  5  6   b

In [6]:
# ..or the last ones
df.tail(2)

a  b  c  d  e  f  g  h  i  j grp
3  2  0  5  0  2  4  5  1  1  7   b
4  2  0  8  0  3  1  0  5  1  1   a

In [7]:
# Select specific columns
df[['d', 'f']]

d  f
0  5  3
1  1  4
2  3  5
3  0  4
4  0  1

In [9]:
# ..rows
df[2:4]

a  b  c  d  e  f  g  h  i  j grp
2  6  3  6  3  8  5  4  6  2  2   a
3  2  0  5  0  2  4  5  1  1  7   b

In [10]:
# Rename columns
df = df.rename(columns = {'a': 'aa', 'b': 'bb'})
list(df.columns)

['aa', 'bb', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'grp']

In [11]:
# Create new column from other columns. Same index, same column operation. 
df['ij'] = df['i'] + df['j']
df[1:3]

aa  bb  c  d  e  f  g  h  i  j grp  ij
1   0   3  5  1  8  4  7  0  5  6   b  11
2   6   3  6  3  8  5  4  6  2  2   a   4

You have more ways to view data.

In [76]:
# Filter specific rows. This reads as - from df, get rows with even values on column f
df[df['f'] % 2 == 0] 

aa  bb  c  d  e  f  g  h  i  j grp  ij
1  0   3   5  1  8  4  7  0  5  6  b   11
3  2   0   5  0  2  4  5  1  1  7  b   8

In [75]:
# Show only wanted columns
df[df['f'] > 5][['f', 'bb']] 

Empty DataFrame
Columns: [f, bb]
Index: []

In [74]:
# Filtering will accept expressions that evaluate as True / False 
df[(df['aa'].isin([3, 5])) | (df['bb'] < 4)][['aa', 'bb']]

aa  bb
0  5   7 
1  0   3 
2  6   3 
3  2   0 
4  2   0

In [15]:
# Group and aggregate!
df['grp'] = pd.DataFrame(['a', 'b'] * 5) # => add a column with discrete values
df.groupby(['grp']).agg({'c': 'sum', 'd': 'mean', 'e': 'min', 'f': 'max'})[['c', 'd', 'e', 'f']]

# => grouped using the categorical vars, then aggregated by column depending on the function specified
# the agg function accepts a dict of {column: func_name or numpy_func}

c         d  e  f
grp                    
a    22  2.666667  0  5
b    10  0.500000  2  4

In [16]:
# ..and order / sort too
df.sort_values(by=['ij'], ascending=[False])[['ij']]

ij
1  11
3   8
0   7
2   4
4   2

In [17]:
# You can chain operations.
df.loc[df['ij'] > 10] \
  .groupby(['grp']) \
  .agg('sum') \
  .sort_values(['ij'], ascending=[True])

aa  bb  c  d  e  f  g  h  i  j  ij
grp                                    
b     0   3  5  1  8  4  7  0  5  6  11

In [43]:
# This is how you join. 
# There are variants (left, right, inner, outer) available.
df_2 = pd.DataFrame([[random.randint(0,9) for i in range(2)] for i in range(5)],
                  index=[i for i in range(5)], 
                  columns=['b_2', 'c_2']) # => another df

df[['bb', 'c']].merge(df_2, left_on='c', right_on='c_2', how='inner')

bb  c  b_2  c_2
0  7   8  2    8  
1  7   8  4    8  
2  0   8  2    8  
3  0   8  4    8  
4  3   6  5    6

Here are some things you will be needing too.

In [19]:
# Append separate datasets w/ same column names
df_3 = pd.DataFrame([[random.randint(0,9) for i in range(10)] for i in range(5)],
                  index=[i for i in range(5)], 
                  columns=list('abcdefghij'))

df.append(df_3).reset_index()[3:7] # => notice NaN values in columns that didn't match

index    a   aa    b   bb  c  d  e  f  g  grp  h  i   ij  j
3      3  NaN  2.0  NaN  0.0  5  0  2  4  5    b  1  1  8.0  7
4      4  NaN  2.0  NaN  0.0  8  0  3  1  0    a  5  1  2.0  1
5      0  4.0  NaN  6.0  NaN  6  3  6  6  6  NaN  6  2  NaN  9
6      1  6.0  NaN  2.0  NaN  0  1  2  7  9  NaN  8  8  NaN  3

In [20]:
# Lambda goooodness! .apply takes each row and performs a function.
df[['bb']].apply(lambda x: x ** 2)

bb
0  49
1   9
2   9
3   0
4   0

In [21]:
# Of course, there's a pivot in there. row values as column values.
df.pivot_table(columns='grp', aggfunc='mean') # => makes each ['grp'] value as columns

grp         a    b
aa   4.333333  1.0
bb   3.333333  1.5
c    7.333333  5.0
d    2.666667  0.5
e    3.666667  5.0
f    3.000000  4.0
g    3.666667  6.0
h    6.666667  0.5
i    2.666667  3.0
ij   4.333333  9.5
j    1.666667  6.5

In [22]:
# A must-have for analysts, describe().. <3
# Basic statistical descriptives
df.describe()

aa        bb         c         d        e         f         g  \
count  5.00000  5.000000  5.000000  5.000000  5.00000  5.000000  5.000000   
mean   3.00000  2.600000  6.400000  1.800000  4.20000  3.400000  4.600000   
std    2.44949  2.880972  1.516575  2.167948  3.63318  1.516575  2.880972   
min    0.00000  0.000000  5.000000  0.000000  0.00000  1.000000  0.000000   
25%    2.00000  0.000000  5.000000  0.000000  2.00000  3.000000  4.000000   
50%    2.00000  3.000000  6.000000  1.000000  3.00000  4.000000  5.000000   
75%    5.00000  3.000000  8.000000  3.000000  8.00000  4.000000  7.000000   
max    6.00000  7.000000  8.000000  5.000000  8.00000  5.000000  7.000000   

              h        i         j         ij  
count  5.000000  5.00000  5.000000   5.000000  
mean   4.200000  2.80000  3.600000   6.400000  
std    3.701351  2.04939  2.701851   3.507136  
min    0.000000  1.00000  1.000000   2.000000  
25%    1.000000  1.00000  2.000000   4.000000  
50%    5.000000  2.00000  2.000000   7.000000  
75%    6.000000  5.00000  6.000000   8.000000  
max    9.000000  5.00000  7.000000  11.000000

That's it. There's a lot more that Pandas can offer. Head to the documentation to check them all out!

Pandas is the hammer, data is the nail. Getting comfortable with the tool means getting hands-on with the things it is supposed to handle.

---

# In Action
Let's use an actual data and do some basic descriptive analysis, while applying new Pandas skills. Along the way, we may learn new stuff. 

Analysis commentary will be in markdown; code implementation will be the python comments.

We'll be using the [Kaggle 2017 Developer Survey](https://www.kaggle.com/kaggle/kaggle-survey-2017/data) dataset to answer a single question - __How popular is Python with the data science community?__ Btw, do check these [kernels](https://www.kaggle.com/kaggle/kaggle-survey-2017/kernels) to see more in-depth analyses of the dataset.

In [23]:
# Load the csv's and store in a dataframe. 
# There were 4 csv's provided, but we'll mainly need these 2 for this part.
schema = pd.read_csv("../resources/kaggle-survey-2017/schema.csv")
responses = pd.read_csv("../resources/kaggle-survey-2017/multipleChoiceResponses.csv", low_memory=False)

In [27]:
pd.set_option('display.max_colwidth', -1) # this just lets you view 'wrapped' cells

# see schema
schema.tail(1)

Column  \
289  JobFactorPublishingOpportunity   

                                                                                   Question  \
289  How are you assessing potential job opportunities? - Opportunity to publish my results   

        Asked  
289  Learners

In [36]:
# a sample row from the answers
responses.tail(1)

GenderSelect Country   Age    EmploymentStatus StudentStatus  \
16715  Male         Japan   27.0  Employed full-time  NaN            

      LearningDataScience CodeWriter CareerSwitcher CurrentJobTitleSelect  \
16715  NaN                 No         Yes            Programmer             

      TitleFit              ...               JobFactorExperienceLevel  \
16715  Fine                 ...                NaN                       

      JobFactorDepartment JobFactorTitle JobFactorCompanyFunding  \
16715  NaN                 NaN            NaN                      

      JobFactorImpact JobFactorRemote JobFactorIndustry  \
16715  NaN             NaN             NaN                

      JobFactorLeaderReputation JobFactorDiversity  \
16715  NaN                       NaN                 

      JobFactorPublishingOpportunity  
16715  NaN                            

[1 rows x 228 columns]

There are 290 questions! There's a lot of data that can be explored, and many questions can be answered. 

For our purposes, let's just trim down this list to the relevant ones, and see where the questions point us at.

In [77]:
# Filter the Question field with the word 'language' - find signal from the noise.
# .str.contains() is a string indexer for the column
schema[schema['Question'].str.contains('language')]

Column  \
18   LanguageRecommendationSelect             
19   LanguageRecommendationFreeForm           
122  WorkToolsSelect                          
123  WorkToolsFreeForm1                       
124  WorkToolsFreeForm2                       
125  WorkToolsFreeForm3                       
126  WorkToolsFrequencyAmazonML               
127  WorkToolsFrequencyAWS                    
128  WorkToolsFrequencyAngoss                 
129  WorkToolsFrequencyC                      
130  WorkToolsFrequencyCloudera               
131  WorkToolsFrequencyDataRobot              
132  WorkToolsFrequencyFlume                  
133  WorkToolsFrequencyGCP                    
134  WorkToolsFrequencyHadoop                 
135  WorkToolsFrequencyIBMCognos              
136  WorkToolsFrequencyIBMSPSSModeler         
137  WorkToolsFrequencyIBMSPSSStatistics      
138  WorkToolsFrequencyIBMWatson              
139  WorkToolsFrequencyImpala                 
140  WorkToolsFrequencyJava                   
141  WorkToolsFrequencyJulia                  
142  WorkToolsFrequencyJupyter                
143  WorkToolsFrequencyKNIMECommercial        
144  WorkToolsFrequencyKNIMEFree              
145  WorkToolsFrequencyMathematica            
146  WorkToolsFrequencyMATLAB                 
147  WorkToolsFrequencyAzure                  
148  WorkToolsFrequencyExcel                  
149  WorkToolsFrequencyMicrosoftRServer       
..                                  ...       
151  WorkToolsFrequencyMinitab                
152  WorkToolsFrequencyNoSQL                  
153  WorkToolsFrequencyOracle                 
154  WorkToolsFrequencyOrange                 
155  WorkToolsFrequencyPerl                   
156  WorkToolsFrequencyPython                 
157  WorkToolsFrequencyQlik                   
158  WorkToolsFrequencyR                      
159  WorkToolsFrequencyRapidMinerCommercial   
160  WorkToolsFrequencyRapidMinerFree         
161  WorkToolsFrequencySalfrod                
162  WorkToolsFrequencySAPBusinessObjects     
163  WorkToolsFrequencySASBase                
164  WorkToolsFrequencySASEnterprise          
165  WorkToolsFrequencySASJMP                 
166  WorkToolsFrequencySpark                  
167  WorkToolsFrequencySQL                    
168  WorkToolsFrequencyStan                   
169  WorkToolsFrequencyStatistica             
170  WorkToolsFrequencyTableau                
171  WorkToolsFrequencyTensorFlow             
172  WorkToolsFrequencyTIBCO                  
173  WorkToolsFrequencyUnix                   
174  WorkToolsFrequencySelect1                
175  WorkToolsFrequencySelect1FreeForm        
176  WorkToolsFrequencySelect2                
177  WorkFrequencySelect2FreeForm             
178  WorkFrequencySelect3                     
179  WorkFrequencySelect3FreeForm             
277  JobFactorLanguages                       

                                                                                                                                                                        Question  \
18   What programming language would you recommend a new data scientist learn first? (Select one option) - Selected Choice                                                         
19   What programming language would you recommend a new data scientist learn first? (Select one option) - Other - Text                                                            
122  For work, which data science/analytics tools, technologies, and languages have you used in the past year? (Select all that apply) - Selected Choice                           
123  For work, which data science/analytics tools, technologies, and languages have you used in the past year? (Select all that apply) - Other - Text1                             
124  For work, which data science/analytics tools, technologies, and languages have you used in the past year? (Select all that apply) - Other - Text2                             
125  For work, which data science/analytic

Let's hone in on the ff: (Question #n will just be the row number; tho the actual is #n + 1)

Question #18: [LanguageRecommendationSelect]    
_What programming language would you recommend a new data scientist learn first?_ will let us know how people recommend Python.

Question #122: [WorkToolsSelect]   
_For work, which data science/analytics tools, technologies, and languages have you used in the past year?_ will let us know the Python users from the respondents.

Question #156:   
_At work, how often did you use the following data science/analytics tools, technologies, and languages this past year? - Python_ will let us know how much Python was used in previous year.

Question #156:   
_How are you assessing potential job opportunities? - The languages, frameworks, and other technologies I'd be working with_ will let us know how Python relates to jobs.

### a) How recommended is Python?
Warming up.

In [71]:
# From the schema, this is the corresponding Column field of the Question field
acol = 'LanguageRecommendationSelect' 

# With a specific column, group specific values, then sort those counts.
responses \
    .groupby(acol) \
    .agg({acol: 'count'}) \
    .sort_values([acol], ascending=[False])

LanguageRecommendationSelect
LanguageRecommendationSelect                              
Python                        6941                        
R                             2643                        
SQL                           385                         
C/C++/C#                      307                         
Matlab                        238                         
Java                          138                         
Scala                         94                          
SAS                           88                          
Other                         85                          
Julia                         30                          
Stata                         28                          
Haskell                       17                          
F#                            4

Most recommended, with a far second and third!    
Recommending something to others is often an indicator of how satisfied the current users are to it. 

### b) Who are the Python users?

Since the top 3 above are Python, R, SQL, let's see how they compare with actual users.

In [175]:
# What do the values look like?
responses[[bcol]].head(5)

# So it's like checkbox type data entry. Multiple values in one answer; comma-separated. 

WorkToolsSelect
0  Amazon Web services,Oracle Data Mining/ Oracle R Enterprise,Perl                                                                                                                                                   
1  NaN                                                                                                                                                                                                                
2  NaN                                                                                                                                                                                                                
3  Amazon Machine Learning,Amazon Web services,Cloudera,Hadoop/Hive/Pig,Impala,Java,Mathematica,MATLAB/Octave,Microsoft Excel Data Mining,Microsoft SQL Server Data Mining,NoSQL,Python,R,SAS Base,SAS JMP,SQL,Tableau
4  C/C++,Jupyter notebooks,MATLAB/Octave,Python,R,TensorFlow

In [167]:
bcol = 'WorkToolsSelect'
total = float(responses[bcol].count())

# .get_dummies will create cols for all possible string values on the comma-separated answer
# The value of each col is either 1 or 0 depending on their appearance on the comma-separated answer.
# e.g "Python,SQL" will have a 1 for Python and SQL cols; 0 for the others
# .sum is for summing all row entries of the columns
# .to_frame to convert to df for display
users = responses[bcol] \
            .str.get_dummies(sep=',') \
            .sum() \
            .to_frame('Count') \
            .sort_values(['Count'], ascending=[False]) \
            [0:10]

# New column with % of the total answers
users['%'] = users['Count'] / total
users

Count         %
Python               6073   0.763419
R                    4708   0.591829
SQL                  4261   0.535638
Jupyter notebooks    3206   0.403017
TensorFlow           2256   0.283595
Amazon Web services  1868   0.234821
Unix shell / awk     1854   0.233061
Tableau              1619   0.203520
C/C++                1528   0.192080
NoSQL                1527   0.191955

The ranking of the 3 languages based on actual users seem to be consistent with the ranking on most recommended.   
Python is mentioned 3 out of 4 times as one of their tools.

Focusing on Python, let's see how the spread looks by age and job title.

In [ ]:
# By age

In [ ]:
# By job title

For non-python users, how much do they like python to recommend it?

# Useful Links
- [Official Pandas Documentation](http://pandas.pydata.org/pandas-docs/stable/)   
- [Pandas Cheatsheet (PDF)](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)   
- [Python for Data Analysis (Book)](https://www.amazon.com/Python-Data-Analysis-Wrangling-IPython/dp/1491957662/ref=pd_sim_14_2?_encoding=UTF8&psc=1&refRID=TJ9Q3J20VHGT3KFN273Q)   
- [Dataschool.io Series on Pandas (Videos)](http://www.dataschool.io/easier-data-analysis-with-pandas/)   
- [Intro to Data Analysis (MOOC)](https://www.udacity.com/course/intro-to-data-analysis--ud170)   